### PR TITLE
SMES and substation UI

### LINKED ISSUE #2878: SMES units really need a UI
See tin.

### PR DESCRIPTION
**Pull request by chromiumboy**
## About the PR

- Added UIs to the SMES and substation. These UIs display the battery charge, current power load, and the quality of the external power supply. The substation also displays all of its sources and loads in a scrollable window
- The external power description includes a new state - 'stable'. Appears when power supply and power demand are roughly equal (within 10% of each other)
- The power monitoring console and APC have also been given a face lift
- SMES, substations and APCs all list their assigned name at the top of their respective UIs
- Duplicated UI code has been removed and is handled by a common system

This PR closes #2878

## Why / Balance

This PR adds missing UI elements to the SMES and substation. The addition of source/load monitoring on the substation will also assist in diagnosing issues in the local power net, which is difficult at present

The power monitor console will continue to serve a role in the game by allowing players to check the status of all SMES and substations on a cable network simultaneously, to get an overview picture of what's going on with the primary storage and distribution system (for further diagnosis, the player would then go to the problematic machine). It will also continue to allow players to monitor the stations primary power status from other places, such as the CE office and bridge

## Technical details

The code for power monitoring has been generalised and is now handled on the server by 'PowerMonitoringSystem' and 'PowerMonitoringComponent'. Client side UI handling for this system (e.g., drawing the charge bars, querying the external power status, updating the lists of sources and loads) have been centralized in the client 'PowerMonitoringSystem'. Messages between the server and client for these systems are contained in 'SharedPowerMonitoringComponent'

'HasAccess' and 'LastExternalState' on ApcComponent is no longer used and have been marked as obsolete

## Media

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

power_uis_v2

## Breaking changes

Power monitoring has been generalized and is no longer restricted to consoles. To mark this transition, some classes have been renamed. To resolve issues relating to this change:

- All files - Replace 'PowerMonitoringConsoleSystem' with 'PowerMonitoringSystem' (client and server)
- All files - Replace 'PowerMonitoringConsoleComponent' with 'PowerMonitoringComponent' (client and server)
- All files - Replace 'SharedPowerMonitoringConsoleComponent' with 'SharedPowerMonitoringComponent'

UI code from the APC is now shared with the SMES and substation. Consequently, some classes have been renamed and methods changed. To resolve issues relating to this change:

- All files - replace 'ApcExternalPowerState' with 'ExternalPowerState'
- ApcSystem - Replace 'CalcExtPowerState(EntityUid uid, PowerState.Battery battery)' with 'PowerMonitoringSystem.CalcExtPowerState(EntityUid uid, PowerState.Battery battery)'
- SharedApcSystem - The arguments for 'ApcBoundInterfaceState' have changed, replace calls to 'ApcBoundInterfaceState(bool mainBreaker, bool hasAccess, int power, ExternalPowerState externalPower, float charge)' with 'ApcBoundInterfaceState(bool mainBreaker, int power, ExternalPowerState externalPower, float charge)'
- ApcMenu.xaml and ApcMenu.xaml.cs - Replace 'BreakerMain' with 'BreakerOnMain'

Some fields on ApcComponent have been marked as obsolete and should not longer be used
- 'HasAccess' - Do not use this field to store access permissions to the APC
- 'LastExternalState' - Use the 'LastExternalState' field on PowerNetworkBatteryComponent instead

**Changelog**

:cl:
- add: UIs have been added for the SMES and substation

### CODE REVIEW

**Comment by Partmedia** [comment]
In "Breaking changes", please add specific instructions for a fork to port over code. If your replacements are drop-in replacements, say that. If other changes (e.g. method names, argument order, etc.) changes, provide a one to one mapping of old functions to new.

**Comment by deltanedas** [comment]
also changelog needs :cl:

File: Content.Client/Power/PowerMonitoring/PowerMonitoringSystem.cs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+using JetBrains.Annotations;
+using Robust.Client.Graphics;
+using Robust.Client.UserInterface.Controls;
+using System.Linq;
+using static Robust.Client.UserInterface.Control;
+
+namespace Content.Client.Power.PowerMonitoring;
+
+[UsedImplicitly]
+public sealed class PowerMonitoringSystem : EntitySystem
```
Comment (AJCM-git) [inline Content.Client/Power/PowerMonitoring/PowerMonitoringSystem.cs:12]: This is super not allowed, system methods shouldnt ever care about UI controls, move these to the window or make a UI controller if necessary

File: Content.Server/Power/Components/PowerMonitoringComponent.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
+using Content.Server.NodeContainer;
+using Content.Server.Power.EntitySystems;
+using Robust.Shared.Reflection;
+using Robust.Shared.Serialization;
+
+namespace Content.Server.Power.Components;
+
+[RegisterComponent, Access(typeof(PowerMonitoringSystem))]
+public sealed partial class PowerMonitoringComponent : Component, ISerializationHooks
```
Comment (AJCM-git) [inline Content.Server/Power/Components/PowerMonitoringComponent.cs:9]: Serialization hooks are deprecated, please dont use them

File: Content.Server/Power/EntitySystems/PowerMonitoringSystem.cs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        // Get loads and sources
+        var totalLoads = GetTotalLoadsForNode(uid, loadNode, out var loads);
+        var totalSources = GetTotalSourcesForNode(uid, sourceNode, out var sources);
+
+        // Sort loads and sources
+        loads.Sort(CompareLoadOrSources);
+        sources.Sort(CompareLoadOrSources);
+
+        // Get battery values (if applicable)
+        Resolve(uid, ref netBattery, false);
```
Comment (AJCM-git) [inline Content.Server/Power/EntitySystems/PowerMonitoringSystem.cs:70]: Either make this a check or use trycomp

File: Content.Server/Power/EntitySystems/PowerMonitoringSystem.cs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+    private double GetTotalSourcesForNode(EntityUid uid, Node node, out List<PowerMonitoringEntry> sources)
+    {
+        var totalSources = 0.0d;
+        sources = new List<PowerMonitoringEntry>();
+
+        if (node.NodeGroup is not PowerNet netQ)
+            return totalSources;
+
+        foreach (PowerSupplierComponent powerSupplier in netQ.Suppliers)
```
Comment (AJCM-git) [inline Content.Server/Power/EntitySystems/PowerMonitoringSystem.cs:89]: Use var (same for the other loops)

File: Content.Server/Power/EntitySystems/PowerMonitoringSystem.cs [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    private void OnBoundUiOpen(EntityUid uid, PowerMonitoringComponent component, BoundUIOpenedEvent args)
+    {
+        UpdateUIState(uid, component);
+    }
+
+    private void OnBatteryChargeChanged(EntityUid uid, PowerMonitoringComponent component, ref ChargeChangedEvent args)
+    {
+        PowerNetworkBatteryComponent? battery = null;
```
Comment (AJCM-git) [inline Content.Server/Power/EntitySystems/PowerMonitoringSystem.cs:192]: Dont use resolves like this, trycomp exists

**Review by AJCM-git — CHANGES_REQUESTED** [review]
.

**Comment by chromiumboy** [comment]
> .

Cheers for the review, I wasn't expecting one so quickly! :)

**Comment by chromiumboy** [comment]
Just FYI, this PR is on hold while I work on improving the power monitoring console itself

**Comment by Partmedia** [comment]
I think these changes are fine, if you'd like to un-draft them and have them in-game as an interim before we get improved power monitors.

**Comment by AJCM-git** [comment]
@chromiumboy you coming back to this?

**Comment by chromiumboy** [comment]
> @chromiumboy you coming back to this?

Not at the moment, and this will need a rethink and major refactoring now that the power monitoring console PR has been merged. So I'll close this for now

**Pull request closed without merging**